thrift.js

```diff
@@ -117,37 +118,112 @@ function Thrift(options) {
 
-    this.filename = options.entryPoint;
+    if (options.entryPoint) {
+        return this.load(options.entryPoint);
+    }
+}
+
+Thrift.prototype.models = 'module';
+
+Thrift.prototype.Message = Message;
+
+Thrift.prototype.load = function load(entryPoint) {
+    this.memo[entryPoint] = this;
+    this.filename = entryPoint;
     this.dirname = path.dirname(this.filename);
-    this.memo[this.filename] = this;
 
-    var ast = this.asts[options.entryPoint];
-    var source = this.idls[options.entryPoint];
-    if (!source && !ast) {
+    this.exception = null;
+
+    if (this.asyncFetch) {
+        var model = this;
+        return this._buildAst(this.filename, this.allowIncludeAlias)
+            .then(function () { model._compileAndLink(); return model; });
+    }
+
+    this._buildAst(this.filename, this.allowIncludeAlias);
+    this._compileAndLink();
+    return this;
+}
+
+Thrift.prototype._buildAst = function _buildAst(filename, allowIncludeAlias, alreadyBuilt) {
+    alreadyBuilt = alreadyBuilt || {};
+    if (alreadyBuilt[filename]) {
+        return Promise.resolve();
+    }
+    alreadyBuilt[filename] = true;
+
+    if (!this.idls[filename] && !this.asts[filename]) {
         /* eslint-disable max-len */
-        assert.ok(this.fs, this.filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
-        assert.ok(this.filename, 'Thrift must be constructed with a options.entryPoint');
+        assert.ok(this.fs, filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
         /* eslint-enable max-len */
-        this.filename = path.resolve(this.filename);
-        source = this.fs.readFileSync(this.filename, 'ascii');
-        this.idls[this.filename] = source;
+        assert.ok(filename, 'Thrift must be passed an entryPoint');
     }
 
-    this.exception = null;
+    if (!this.asts[filename] && this.idls[filename]) {
+        this.asts[filename] = idl.parse(this.idls[filename]);
+    }
+    if (!this.asts[filename]) {
+        if (this.asyncFetch) {
+            var model = this;
+            return this.fs.load(filename)
+                .then(function (source) {
+                    model.idls[filename] = source;
+                    model.asts[filename] = idl.parse(source);
+                    return model._buildChildAsts(filename, model.asts[filename], allowIncludeAlias, alreadyBuilt);
+                });
+        } else {
+            this.idls[filename] = this.fs.readFileSync(path.resolve(filename), 'ascii');
+            this.asts[filename] = idl.parse(this.idls[filename]);
+        }
+    }
+
+    this._buildChildAsts(filename, this.asts[filename], allowIncludeAlias, alreadyBuilt);
+    return Promise.resolve();
+}
+
+Thrift.prototype._buildChildAsts = function _buildChildAsts(filename, syntax, allowIncludeAlias, alreadyBuilt) {
```

Comment:
nit: “build AST” is a long way to say “parse”.

Perhaps `_parseIncludedModules`?